File: output/xCodeEval_remove_8000/C++/C/485dc9447645f5f29b8cb9a8c04b832e.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <stdint.h>

#define MAX_N 2000
#define MAX_K 2000

int64_t max(int64_t a, int64_t b) {
    return (a > b) ? a : b;
}

void add_item(int64_t *knapsack, int size, int64_t sum, int K) {
    for (int k = K - size; k >= 0; k--) {
        knapsack[k + size] = max(knapsack[k + size], knapsack[k] + sum);
    }
}

void recurse(int start, int end, int64_t *knapsack, int N, int K, int64_t **A, int64_t *A_sum, int64_t *best) {
    if (start >= end) {
        return;
    }

    if (end - start == 1) {
        int64_t sum = 0;
        *best = max(*best, knapsack[K]);

        for (int prefix = 1; prefix <= A[start][0]; prefix++) {
            sum += A[start][prefix];
            *best = max(*best, sum + knapsack[K - prefix]);
        }

        return;
    }

    int mid = (start + end) / 2;
    int64_t *state = malloc((K + 1) * sizeof(int64_t));
    for (int i = 0; i <= K; i++) {
        state[i] = knapsack[i];
    }

    for (int i = start; i < mid; i++) {
        add_item(state, A[i][0], A_sum[i], K);
    }

    recurse(mid, end, state, N, K, A, A_sum, best);
    free(state);

    state = malloc((K + 1) * sizeof(int64_t));
    for (int i = 0; i <= K; i++) {
        state[i] = knapsack[i];
    }

    for (int i = mid; i < end; i++) {
        add_item(state, A[i][0], A_sum[i], K);
    }

    recurse(start, mid, state, N, K, A, A_sum, best);
    free(state);
}

void solve() {
    int N, K;
    scanf("%d %d", &N, &K);

    int64_t **A = malloc(N * sizeof(int64_t *));
    int64_t *A_sum = malloc(N * sizeof(int64_t));

    for (int i = 0; i < N; i++) {
        int t;
        scanf("%d", &t);
        A[i] = malloc((t + 1) * sizeof(int64_t));
        A[i][0] = t;

        for (int j = 1; j <= t; j++) {
            scanf("%lld", &A[i][j]);
        }

        if (t > K) {
            A[i][0] = K;
        }

        A_sum[i] = 0;
        for (int j = 1; j <= A[i][0]; j++) {
            A_sum[i] += A[i][j];
        }
    }

    int64_t best = 0;
    int64_t *knapsack = malloc((K + 1) * sizeof(int64_t));
    for (int i = 0; i <= K; i++) {
        knapsack[i] = 0;
    }

    recurse(0, N, knapsack, N, K, A, A_sum, &best);

    printf("%lld\n", best);

    free(knapsack);
    for (int i = 0; i < N; i++) {
        free(A[i]);
    }
    free(A);
    free(A_sum);
}

int main() {
    solve();
    return 0;
}

File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/Python/485dc9447645f5f29b8cb9a8c04b832e.py
import sys
from typing import List, Tuple

def add_item(knapsack: List[int], size: int, sum: int) -> None:
    for k in range(K - size, -1, -1):
        knapsack[k + size] = max(knapsack[k + size], knapsack[k] + sum)

def recurse(start: int, end: int, knapsack: List[int]) -> None:
    if start >= end:
        return
    
    if end - start == 1:
        sum = 0
        global best
        best = max(best, knapsack[K])
        
        for prefix in range(1, len(A[start]) + 1):
            sum += A[start][prefix - 1]
            best = max(best, sum + knapsack[K - prefix])
        
        return
    
    mid = (start + end) // 2
    state = knapsack.copy()
    
    for i in range(start, mid):
        add_item(state, len(A[i]), A_sum[i])
    
    recurse(mid, end, state)
    state = knapsack.copy()
    
    for i in range(mid, end):
        add_item(state, len(A[i]), A_sum[i])
    
    recurse(start, mid, state)

def solve(N: int, K: int, A: List[List[int]]) -> int:
    global best
    A_sum = [sum(a) for a in A]
    best = 0
    knapsack = [0] * (K + 1)
    recurse(0, N, knapsack)
    return best

def from_input_string(input_string: str) -> Tuple[int, int, List[List[int]]]:
    lines = input_string.strip().split('\n')
    N, K = map(int, lines[0].split())
    A = [list(map(int, line.split())) for line in lines[1:]]
    return N, K, A

def to_input_string(inputs: Tuple[int, int, List[List[int]]]) -> str:
    N, K, A = inputs
    res = []
    res.append(f"{N} {K}")
    for a in A:
        res.append(' '.join(str(x) for x in a))
    return '\n'.join(res)

def from_output_string(output_string: str) -> int:
    return int(output_string.strip())

def to_output_string(output: int) -> str:
    return str(output)
    
